benchmark/benchmain/main.go

```diff
@@ -169,76 +175,92 @@ func runModesFromWorkloads(workload string) runModes {
 	return r
 }
 
-func unaryBenchmark(startTimer func(), stopTimer func(uint64), benchFeatures stats.Features, benchTime time.Duration, s *stats.Stats) uint64 {
-	caller, cleanup := makeFuncUnary(benchFeatures)
+type startFunc func()
+type stopFunc func(count uint64)
+type ucStopFunc func(req uint64, resp uint64)
+type rpcCallFunc func(pos int)
+type rpcSendFunc func(pos int)
+type rpcRecvFunc func(pos int)
+type rpcCleanupFunc func()
+
+func unaryBenchmark(start startFunc, stop stopFunc, bf stats.Features, s *stats.Stats) {
+	caller, cleanup := makeFuncUnary(bf)
 	defer cleanup()
-	return runBenchmark(caller, startTimer, stopTimer, benchFeatures, benchTime, s)
+	runBenchmark(caller, start, stop, bf, s)
 }
 
-func streamBenchmark(startTimer func(), stopTimer func(uint64), benchFeatures stats.Features, benchTime time.Duration, s *stats.Stats) uint64 {
-	caller, cleanup := makeFuncStream(benchFeatures)
+func streamBenchmark(start startFunc, stop stopFunc, bf stats.Features, s *stats.Stats) {
+	caller, cleanup := makeFuncStream(bf)
 	defer cleanup()
-	return runBenchmark(caller, startTimer, stopTimer, benchFeatures, benchTime, s)
+	runBenchmark(caller, start, stop, bf, s)
 }
 
-func unconstrainedStreamBenchmark(benchFeatures stats.Features, warmuptime, benchTime time.Duration) (uint64, uint64) {
-	var sender, recver func(int)
-	var cleanup func()
-	if benchFeatures.EnablePreloader {
-		sender, recver, cleanup = makeFuncUnconstrainedStreamPreloaded(benchFeatures)
+func unconstrainedStreamBenchmark(start startFunc, stop ucStopFunc, bf stats.Features, s *stats.Stats) {
+	var sender rpcSendFunc
+	var recver rpcRecvFunc
+	var cleanup rpcCleanupFunc
+	if bf.EnablePreloader {
+		sender, recver, cleanup = makeFuncUnconstrainedStreamPreloaded(bf)
 	} else {
-		sender, recver, cleanup = makeFuncUnconstrainedStream(benchFeatures)
+		sender, recver, cleanup = makeFuncUnconstrainedStream(bf)
 	}
 	defer cleanup()
 
-	var (
-		wg            sync.WaitGroup
-		requestCount  uint64
-		responseCount uint64
-	)
-	wg.Add(2 * benchFeatures.MaxConcurrentCalls)
-
-	// Resets the counters once warmed up
+	var req, resp uint64
 	go func() {
+		// Resets the counters once warmed up
 		<-time.NewTimer(warmuptime).C
-		atomic.StoreUint64(&requestCount, 0)
-		atomic.StoreUint64(&responseCount, 0)
+		atomic.StoreUint64(&req, 0)
+		atomic.StoreUint64(&resp, 0)
+		start()
 	}()
 
-	bmEnd := time.Now().Add(benchTime + warmuptime)
-	for i := 0; i < benchFeatures.MaxConcurrentCalls; i++ {
+	bmEnd := time.Now().Add(bf.BenchTime + warmuptime)
+	var wg sync.WaitGroup
+	wg.Add(2 * bf.MaxConcurrentCalls)
+	for i := 0; i < bf.MaxConcurrentCalls; i++ {
 		go func(pos int) {
+			defer wg.Done()
 			for {
 				t := time.Now()
 				if t.After(bmEnd) {
-					break
+					return
 				}
+				start := time.Now()
 				sender(pos)
-				atomic.AddUint64(&requestCount, 1)
+				elapse := time.Since(start)
+				atomic.AddUint64(&req, 1)
+				s.AddReqDuration(elapse)
```

Comment:
s.AddReqDuration waits on a lock. This function is called concurrently bf.MaxConcurrentCalls times. So, this may create a lock contention and then we will be measuring the performance of lock acquisition and release, instead of RPCs.

Reply:
You are right. But our throughput and QPS numbers are measuring the number of requests and/or responses in a fixed amount of time and these numbers would be reduced due to lock acquisition and release. In the unconstrained streaming mode, I think having accurate throughput numbers is more important than measuring the latencies.